### PR TITLE
Addition of pixel Tracking in HI pp_on_AA_2018 era

### PR DESCRIPTION
**Pull request by abaty**
This PR creates a new era which is a copy of Run2_2018_pp_on_AA, and adds a pixel tracking sequence to the reconstruction.  This pixel tracking sequence has been requested to meet physics goals requiring low-pt tracks with a higher purity than what is achievable with full tracks.  The additional pixel tracking modules have been found to have a negligible impact on the total timing.  Storing the extra track collection has been estimated to increase the event content size of AOD by 11%, which is a number the HI PAG is willing to accept.  Details of the pixel tracking performance in MB HI events can be found at [1].

The PR can be tested with the normal pp_on_AA_2018 cmsDriver command with the era changed to the 'Run2_2018_pp_on_AA_wPixelTrk'.

[1] https://indico.cern.ch/event/749011/contributions/3098671/attachments/1698408/2736177/pixelUpdate_20180727.pdf

@mandrenguyen

### CODE REVIEW

**Comment by slava77** [comment]
> The additional pixel tracking modules have been found to have a negligible impact on the total timing.

Why do we then need a new era?
can this new collection be added to the already existing pp_on_AA_2018 ?

**Comment by abaty** [comment]
My understanding is that depending on the final data-taking strategy, there could be a request to only run the pixel tracking on a subset of all the PDs gathered for the HI run in 2018.  In order to facilitate running two different sequences on specific PDs, I was told it would be beneficial to keep the pixel tracking as a separate era.  Let me know if this is not the case. @slava77

**Comment by slava77** [comment]
On 8/9/18 7:03 AM, abaty wrote:
> My understanding is that depending on the final data-taking strategy,
> there could be a request to only run the pixel tracking on a subset of
> all the PDs gathered for the HI run in 2018. In order to facilitate
> running two different sequences on specific PDs, I was told it would be
> beneficial to keep the pixel tracking as a separate era. Let me know if
> this is not the case. @slava77

ok,
from your most recent response it sounds like the statement in the PR
description
"Storing the extra track collection has been estimated to increase the
event content size of AOD by 11%, which is a number the HI PAG is
willing to accept."
is a bit of wishful thinking.

BTW, how does this increment look like for minbias or for something very
central?

**Comment by mandrenguyen** [comment]
@slava77 Based on the decision at the last HIN PAG meeting, we are indeed willing to add 11% to AOD.   Regarding the two eras, this was requested for the minbias datasets, so I suggested to make a separate era.   This being said, I don't think it was  yet excluded that we just store pixel tracks for all PDs.   Let us get clarification at tomorrow's PAG meeting and we can update this PR if necessary.

**Comment by abaty** [comment]
We recently checked the size increase in b=0 events, and it looks like the number is still 11%.

File: RecoHI/HiTracking/python/HILowPtConformalPixelTracks_cfi.py [hunk 1]
```diff
@@ -41,6 +41,89 @@
...
+	doublets = "hiConformalPixelTracksPhase1HitDoubletsCA",
+        CAPhiCut = cms.double(0.2),
+        CAThetaCut = cms.double(0.0012),
+	SeedComparitorPSet = cms.PSet( 
+           ComponentName = cms.string('none')
+        ),
+        extraHitRPhitolerance = cms.double(0.032),
+        maxChi2 = cms.PSet(
+           enabled = cms.bool(True),
+           pt1 = cms.double(0.7),
```
Comment (slava77) [inline RecoHI/HiTracking/python/HILowPtConformalPixelTracks_cfi.py]: please drop type specifications for all parameters which already exist.
This is a safer way to protect from parameter name mistakes and will also help in possible parameter migrations.

File: RecoHI/HiTracking/python/HILowPtConformalPixelTracks_cfi.py [hunk 2]
```diff
@@ -41,6 +41,89 @@
...
+        lipMax = cms.double(999.0),
+        nSigmaLipMaxTolerance = cms.double(999.9),
+        nSigmaTipMaxTolerance = cms.double(999.0),
+        ptMax = cms.double(999999),
+        ptMin = cms.double(0.30),
+        tipMax = cms.double(999.0)
+)
+
+#Pixel tracks
+hiConformalPixelTracksPhase1 = cms.EDProducer("PixelTrackProducer",
```
Comment (slava77) [inline RecoHI/HiTracking/python/HILowPtConformalPixelTracks_cfi.py]: I think that "Phase1" can be dropped from the producer name.

**Comment by slava77** [comment]
@mandrenguyen 
please clarify when you expect to have a decision on this.
Thank you.

**Comment by abaty** [comment]
Based on the HI PAG meeting on Friday, the decision was made to move the pixel tracking into the main pp_on_AA_2018 era.  To try to mitigate some of the event size increase, we will also slightly bump up the low-pt threshholds of the generalTracks (as these tracks are mostly fake anyways).

I'm finishing up some tests on the new setup and should push is soon.

**Comment by abaty** [comment]
I've moved the changes in this PR back into the main pp_on_AA_2018 era, as discussed in my previous comment.  The pt threshholds for the low pt steps have been raised from 0.3 to 0.49 GeV as well.  This reduces the number of tracks under 0.5 GeV stored, reducing the extra event content added to AOD in this PR from 11% down to around 7%.  RECO now also runs slightly faster because of the timing saved in the low-pt step.

I've also made changes addressing comments from Slava's code review.

**Comment by slava77** [comment]
@cmsbuild please test workflow 158.0

**Comment by slava77** [comment]
@abaty 
please edit the PR title and description to match the latest implementation.
There is no new era/modifier.

File: RecoHI/HiTracking/python/HILowPtConformalPixelTracks_cfi.py [hunk 3]
```diff
@@ -103,22 +106,22 @@
...
 )
 
-#Pixel tracks
-hiConformalPixelTracksPhase1 = cms.EDProducer("PixelTrackProducer",
-    Cleaner = cms.string('pixelTrackCleanerBySharedHits'),
-    Filter = cms.InputTag("hiConformalPixelTracksPhase1Filter"),
-    Fitter = cms.InputTag("pixelFitterByConformalMappingAndLine"),
-    SeedingHitSets = cms.InputTag("hiConformalPixelTracksPhase1HitQuadrupletsCA"),
+from Configuration.Eras.Modifier_pp_on_AA_2018_cff import pp_on_AA_2018
+pp_on_AA_2018.toModify(hiConformalPixelTracks,
```
Comment (slava77) [inline RecoHI/HiTracking/python/HILowPtConformalPixelTracks_cfi.py]: shouldn't this be ```phase1Pixel.toModify```?

**Comment by slava77** [comment]
Here are some observations to go with the PR

Using 230 minbias events after the filter "3000":
there is about 10% drop in the number of generalTracks
all_sign1034vsorig_hinpbpbmbfilter3000c_log10recotracks_generaltracks__reco_obj_pt
mostly from low pt.

all_sign1034vsorig_hinpbpbmbfilter3000c_recotracks_generaltracks__reco_obj_algo
low-pt triplet and quad iterations are down, in line with the increased thresholds, but the later iterations are now picking up more tracks (including pixel pair).

particle flow follows with a drop at very low pt and some increase up to 1.5 GeV, from the later iterations; overall 12% fewer charged candidates
all_sign1034vsorig_hinpbpbmbfilter3000c_log10recopfcandidates_particleflow__reco_obj_pt46

the pixel tracks show up as expected. There are about 5K tracks per event (only about 20% less than in the generalTracks collection)
pixeltrack_origvssign1034_hinpbpbmbfilter3000c_log10recotracks_hiconformalpixeltracks__reco_obj_pt


CPU use decreased by about 2%:
- the tracking time total is roughly unchanged : a reduction in the lowPt iterations is balanced out by the added pixel tracks and also increase in time spent on later iterations.
- overall decrease in the number of tracks and PF charged candidates reduces time in various high level reco, which adds up to about 2% and explains a total CPU reduction

The event size is up by about 7.5% in AODSIM output
- hiConformalPixelTracks contribute an increase of about 11%, as expected from the PR description
- the reduction in generalTracks and pf candidates dominates the overall reduction and adds up to 2.3%

Memory peak in a production-like job with 8 threads increased by 1 GB (from 15 GB to 16 GB). This is apparently the cost of adding the pixel tracks, but is perhaps also partly coming from the increase in iterative tracking seeds and candidates for higher iterations.

Overall, the results appear to be somewhat as expected.

**Comment by slava77** [comment]
similar plots for wf 158.0 (200 events):

all_sign1034vsorig_hydjetqb12pprecoin2018wf158p0c_log10recotracks_generaltracks__reco_obj_pt

all_sign1034vsorig_hydjetqb12pprecoin2018wf158p0c_recotracks_generaltracks__reco_obj_algo

all_sign1034vsorig_hydjetqb12pprecoin2018wf158p0c_log10recopfcandidates_particleflow__reco_obj_pt46

there is essentially no visible increase in the number of tracks with pt around 1 GeV. Here the fraction of fakes is apparently much smaller.

The change in the fake rate can be seen from the MTV plots:
- for the wf 158.0 it did not increase much and remains low
wf158 0_general09_fr
- for the "filter 3000" events it has gone up significantly
wffilter3000_general09_fr

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/24257	68367e8
- code [configuration] changes are in line with the (updated) PR description and the follow up notes in https://github.com/cms-sw/cmssw/pull/24257#issuecomment-412641054 (the increase in iter tracking thresholds ). The changes affect only the ```pp_on_AA_2018``` modifiers.
- jenkins tests pass and comparisons with the baseline show no differences (the ```pp_on_AA_2018``` is not in the standard tests).
- local tests with wf 158.0 and "filter 3000" events show somewhat expected behavior (see notes in https://github.com/cms-sw/cmssw/pull/24257#issuecomment-414463507 and  https://github.com/cms-sw/cmssw/pull/24257/#issuecomment-414469228)

**Comment by kpedro88** [comment]
+1

**Comment by kpedro88** [comment]
merge

**Pull request merged**